doc/design/logging.md

```diff
@@ -0,0 +1,385 @@
+   shall be possible to replace the default logger at runtime and delegate the
+   log messages to the logger of such a framework. To prevent misuse, this can
+   only be done during the initialization phase. Afterwards an error message shall
+   be logged in the new and old logger and the error handler shall be called with
+   a moderate error level.
+
+7. The default logger shall be replaceable via the platform abstraction in order
+   to use the dedicated logger for log messages which are emitted before it could
+   be replaced at runtime, e.g. from global objects before `main`.
+
+8. Additionally, there shall be a compile time option to forward all log messages
+   independent of the log level threshold for cases where the logger framework
+   does the filtering itself.
+
+9. For tests, the default logger shall be replaced to suppress the log messages in
+   passed tests but output them for failed tests. This keeps the noise low but
+   still provides useful information when needed. Additionally, there shall be an
+   environment variable to circumvent the suppression when desired.
+
+### Solution
+
+#### Class diagram
+
+![logging class diagram](../website/images/logging_classes.svg)
+
+The logger can be customized at compile time and at runtime. The former is done
+by the `BaseLogger` template parameter and the latter by deriving from the logger.
+
+#### Logging with LogStream
+
+![logging with logstream](../website/images/logging_with_logstream.svg)
+
+#### Logging macro
+
+The `IOX_LOG` macro is intended for general use. It has one parameter which
+sets the log level and delegates `file`, `line` and `function` to the `IOX_LOG_INTERNAL`
+macro. The latter can be used in places like a custom `ASSERT` macro which already
+obtained the `file`, `line` and `function` parameter.
+
+The `IOX_LOG_INTERNAL` uses the `iox::log::internal::SelectedLogStream` struct which is
+specialized for `LogLevel::OFF` to compile to a no-op when the logging data does not
+cause side effects. `IOX_LOG(INFO) << 42;` won't cause side effects but
+`IOX_LOG(INFO) << functionCall();` might have some.
+
+The `IOX_LOG_INTERNAL` calls `self()` on the `LogStream` instance to create an lvalue
+reference to the `LogStream` instance. This eases the implementation of logging
+support for custom types since `IOX_LOG(INFO) << myType;` would require to implement
+an overload with a rvalue `LogStream` reference but `IOX_LOG(INFO) << "#### " << myType;`
+requires a lvalue reference.
+
+#### Behavior before calling Logger::init
+
+In order to have log messages before `Logger::init` is called, the default logger
+is used with `LogLevel::INFO`. It is up to the implementation of the default
+logger what to do with these messages. For iceoryx the default logger is the
+`ConsoleLogger` (this can be changed via the platform abstraction) which will
+print the log messages to the console.
+
+Although, it is possible to use the logger without calling `Logger::init`, this is
```

Comment:
What about the use case of a global static class which is using the logger. When the program terminates it is possible that after main is finished the logger goes out of scope before the other global static class does and then we have a segfault or undefined behavior.

Would it makes sense to add this as a restriction for now and implement a nifty counter approach for the logger later to solve this issue.

Reply:
The nifty counter will be part of the `PolymorphicHandler` and we will get it for free once we use it. But that only helps you once you know that you have a problem. Have to think where it makes sense to add such an information.